integrations/tensorflow/iree_tf_compiler/TFL/test/lower_global_tensors.mlir

```diff
@@ -100,6 +100,36 @@ module {
 
 // -----
 
+// CHECK-LABEL: module {
+module {
+  // func @readAssignQuant
+  func @readAssignQuant(%arg0: tensor<2x3x!quant.uniform<i8:f32, 0.1:2>>) -> (tensor<2x3x!quant.uniform<i8:f32, 0.1:2>>) {
+    "tfl.call_once"() {session_init_function = "ReadAssignInit"} : () -> ()
+  // CHECK: util.global private mutable @__iree_flow_Variable = dense<42> : tensor<2x3xi8>
+    %0 = "tfl.var_handle"() {container = "", shared_name = "Variable"} : () -> tensor<*x!tf_type.resource>
+
+    // CHECK: %[[ADDR:.+]] = util.global.load.indirect %ptr___iree_flow_Variable : !util.ptr<tensor<2x3xi8>> -> tensor<2x3xi8>
+    // CHECK: %[[CAST:.+]] = builtin.unrealized_conversion_cast %[[ADDR]] : tensor<2x3xi8> to tensor<2x3x!quant.uniform<i8:f32, 1.000000e-01:2>>
+    %1 = "tfl.read_variable"(%0) : (tensor<*x!tf_type.resource>) -> tensor<2x3x!quant.uniform<i8:f32, 0.1:2>>
+
+    // CHECK: %[[ADD:.+]] = tfl.add %[[CAST]], %arg0 {fused_activation_function = "NONE"}
+    %2 = tfl.add %1, %arg0 {fused_activation_function = "NONE"} : tensor<2x3x!quant.uniform<i8:f32, 0.1:2>>
+
+    // CHECK: %[[CAST2:.+]] = builtin.unrealized_conversion_cast %[[ADD]] : tensor<2x3x!quant.uniform<i8:f32, 1.000000e-01:2>> to tensor<2x3xi8>
```

Comment:
Yeah this worries me a bit. We should be careful about letting these casts bleed between passes and if we do it should be until some finite later point in the pipeline. In the MHLO lowering to linalg we have these that survive a pass because it's a two-pass lowering, but then we do a pass to make sure they're cleaned up and hard-fail if they aren't. If these make it too far down then everything gets real messed up. https://github.com/google/iree/blob/bd2001aa8405373e5a06bf431e1b8b7459a5ddc0/iree/compiler/InputConversion/MHLO/Passes.cpp#L63-L69

At the very least we should make sure it doesn't make it past here https://github.com/google/iree/blob/bd2001aa8405373e5a06bf431e1b8b7459a5ddc0/integrations/tensorflow/iree_tf_compiler/TF/Passes.cpp#L111, but currently that doesn't prevent them: https://github.com/google/iree/blob/bd2001aa8405373e5a06bf431e1b8b7459a5ddc0/integrations/tensorflow/iree_tf_compiler/TF/VerifyFullyConverted.cpp#L39

So at least we should make it illegal there, but I think in addition we should probably throw a reconcile-unrealized-conversion-cast after this pass or whatever is expected to clean this up.

Reply:
It should be resolved by quant type stripping and canonicalizer. Included a check that it is gone once all the TFLite ops are. Also added the reconcile pass prior. Should be safe given how TFLite uses quantized types.